openembedded-core-mbl/scripts/lib/mbl-signing/signing/pki/x509utils.py

```diff
@@ -0,0 +1,315 @@
+        :params critical bool: Specifies whether to mark this
+        extension as critical or not.
+        :params value bytes: Extension value in ASN1 format.
+        """
+        self._oid = x509.ObjectIdentifier(oid)
+        self._critical = critical
+        self._value = x509.UnrecognizedExtension(self.oid, value)
+
+    def __str__(self):
+        """Return a string representation of the object."""
+        return "{};{};{}".format(
+            self.oid.dotted_string,
+            str(self.critical).lower(),
+            binascii.hexlify(self.value.value).decode(),
+        )
+
+    @property
+    def oid(self):
+        """Object Identifier."""
+        return self._oid
+
+    @property
+    def critical(self):
+        """Specify if the extension will be marked as critical."""
+        return self._critical
+
+    @property
+    def value(self):
+        """Extension value wrapped in an UnrecognizedExtension."""
+        return self._value
+
+    @value.setter
+    def value(self, new_val):
+        """
+        Extension value setter.
+
+        :params new_val bytes: New value for the extension.
+        """
+        self._value = x509.UnrecognizedExtension(self.oid, new_val)
+
+    @abstractmethod
+    def encode_value(self, value_data):
+        """Subclass must override this to encode specific value types."""
+        pass
+
+
+class HashExtension(ATFExtension):
+    """An ATF extension containing a hash of a boot component."""
+
+    def __init__(self, *args, content_file=None, **kwargs):
+        """
+        Initialise the extension object.
+
+        :params content_file str: Name of the content file to hash.
+        """
+        super().__init__(*args, **kwargs)
+        self.content_file = content_file
+
+    def encode_value(self, data):
+        """
+        Create a digest of content_file and encode to an ASN1 value.
+
+        :params data bytes: File data in bytes.
+        """
+        if data is None:
+            digest = b"\x00" * 32
+        else:
+            hasher = hashes.Hash(hashes.SHA256(), backend=default_backend())
+            hasher.update(data)
+            digest = hasher.finalize()
+        enc_val = asn1parse.encode_asn1_digest_info(
+            "2.16.840.1.101.3.4.2.1", digest
+        )
+        self.value = enc_val
+
+
+class PKExtension(ATFExtension):
+    """ATFExtension containing a SubjectPublicKeyInfo value."""
+
+    def __init__(self, *args, pub_key_name=None, **kwargs):
+        """
+        Init the extension object.
+
+        :params pub_key_name str: The name of the public key to
+        include in the extension.
+        """
+        super().__init__(*args, **kwargs)
+        self.pub_key_name = pub_key_name
+
+    def encode_value(self, pub_key):
+        """
+        Encode a subject public key and store in the extension.
+
+        :params pub_key bytes: The raw subject public key bytes.
+        """
+        enc_val = asn1parse.encode_asn1_subject_key_info(
```

Comment:
do we care if pub_key is None?

Reply:
`pub_key` is never `None` here, that's handled in the `ExtensionValueGetter`. If `pub_key` is `None` it returns \x00*32. This isn't going to change for the BETA.